pgmpy/factors/Factor.py

```diff
@@ -496,11 +496,24 @@ def __eq__(self, other):
         if type(self) != type(other):
             return False
 
-        if self.variables == other.variables and all(self.cardinality == other.cardinality) \
-                and all(self.values == other.values):
-            return True
-        else:
+        if set(self.scope()) != set(other.scope()):
             return False
+        else:
+            order_dict = {}
+            for var_index in range(len(self.scope())):
+                order_dict[var_index] = other.scope().index(self.scope()[var_index])
+            if any(self.cardinality[[order_dict[i] for i in range(len(self.cardinality))]] != other.cardinality):
+                return False
+            indexes = np.array(list(itertools.product(*[range(i) for i in self.cardinality])))
```

Comment:
I am not very supportive of converting iterator to ```list``` and then converting to ```np.array```. Have a look at ```np.fromiter```. Converting to ```list``` would not be very scalable (I think for large factors it would suck the whole memory). Your thoughts?